clang/lib/Index/IndexRecordHasher.cpp

```diff
@@ -14,173 +14,185 @@
 #include "clang/AST/DeclVisitor.h"
 #include "clang/Basic/FileManager.h"
 #include "clang/Basic/SourceManager.h"
+#include "llvm/Support/BLAKE3.h"
+#include "llvm/Support/HashBuilder.h"
 #include "llvm/Support/Path.h"
 
-#define INITIAL_HASH 5381
-#define COMBINE_HASH(...) (Hash = hash_combine(Hash, __VA_ARGS__))
-
 using namespace clang;
 using namespace clang::index;
 using namespace llvm;
 
-static hash_code computeHash(const TemplateArgument &Arg,
-                             IndexRecordHasher &Hasher);
-static hash_code hashLoc(const SourceManager &SM, SourceLocation Loc,
-                         bool IncludeOffset);
-
 namespace {
-class DeclHashVisitor : public ConstDeclVisitor<DeclHashVisitor, hash_code> {
+
+struct IndexRecordHasher {
+  ASTContext &Ctx;
+  llvm::HashBuilder<llvm::TruncatedBLAKE3<8>, llvm::endianness::little>
+      HashBuilder;
+
+  explicit IndexRecordHasher(ASTContext &context) : Ctx(context) {}
+
+  void hashDecl(const Decl *D);
+  void hashMacro(const IdentifierInfo *name, const MacroInfo *macroInfo);
+
+  void hashType(QualType ty);
+  void hashCanonicalType(CanQualType canTy);
+  void hashDeclName(DeclarationName name);
+  void hashNameSpec(const NestedNameSpecifier *nestedName);
+  void hashSelector(Selector selector);
+  void hashTemplateName(TemplateName name);
+  void hashTemplateArg(const TemplateArgument &arg);
+  void hashLoc(SourceLocation loc, bool includeOffset);
+  void hashAPInt(const APInt &val);
+};
+
+class DeclHashVisitor : public ConstDeclVisitor<DeclHashVisitor, void> {
   IndexRecordHasher &Hasher;
 
 public:
   DeclHashVisitor(IndexRecordHasher &Hasher) : Hasher(Hasher) {}
 
-  hash_code VisitDecl(const Decl *D) {
-    return VisitDeclContext(D->getDeclContext());
+  void VisitDecl(const Decl *decl) {
```

Comment:
What’s the current LLVM naming convention regarding uppercase variables names? I know that we’re trying to move to lowercase variable names in the Swift compiler but I’m not sure about LLVM.

Reply:
Not sure honestly. https://discourse.llvm.org/t/top-level-clang-tidy-options-and-variablename-suggestion-on-codingstandards/58783 was the latest discussion I could find and anecdotally I see a bunch of camelCase in `clang` now.